src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,822 @@
+                }
+
+                if (impersonationLevel == TokenImpersonationLevel.Delegation)
+                {
+                    flags |= ContextFlagsPal.Delegate;
+                }
+            }
+
+            _canRetryAuthentication = false;
+
+            try
+            {
+                _context = new NTAuthentication(isServer, package, credential, servicePrincipalName, flags, channelBinding);
+            }
+            catch (Exception e)
+            {
+                throw new AuthenticationException(SR.net_auth_SSPI, e);
+            }
+        }
+
+        private Exception SetException(Exception e)
+        {
+            if (_exception == null || !(_exception is ObjectDisposedException))
+            {
+                _exception = e;
+            }
+
+            if (_exception != null && _context != null)
+            {
+                _context.CloseContext();
+            }
+
+            return _exception;
+        }
+
+        internal bool IsAuthenticated
+        {
+            get
+            {
+                return _context != null && HandshakeComplete && _exception == null && _remoteOk;
+            }
+        }
+
+        internal bool IsMutuallyAuthenticated
+        {
+            get
+            {
+                if (!IsAuthenticated)
+                {
+                    return false;
+                }
+
+                // Suppressing for NTLM since SSPI does not return correct value in the context flags.
+                if (_context.IsNTLM)
+                {
+                    return false;
+                }
+
+                return _context.IsMutualAuthFlag;
+            }
+        }
+
+        internal bool IsEncrypted
+        {
+            get
+            {
+                return IsAuthenticated && _context.IsConfidentialityFlag;
+            }
+        }
+
+        internal bool IsSigned
+        {
+            get
+            {
+                return IsAuthenticated && (_context.IsIntegrityFlag || _context.IsConfidentialityFlag);
+            }
+        }
+
+        internal bool IsServer
+        {
+            get
+            {
+                return _context != null && _context.IsServer;
+            }
+        }
+
+        internal bool CanGetSecureStream
+        {
+            get
+            {
+                return (_context.IsConfidentialityFlag || _context.IsIntegrityFlag);
+            }
+        }
+
+        internal TokenImpersonationLevel AllowedImpersonation
+        {
+            get
+            {
+                CheckThrow(true);
+                return PrivateImpersonationLevel;
+            }
+        }
+
+        private TokenImpersonationLevel PrivateImpersonationLevel
+        {
+            get
+            {
+                // We should suppress the delegate flag in NTLM case.
+                return (_context.IsDelegationFlag && _context.ProtocolName != NegotiationInfoClass.NTLM) ? TokenImpersonationLevel.Delegation
```

Comment:
A) Check _context for null.
B) Can this be rewritten to not be a ternary expression within a ternary expression?  It's... a bit hard to parse.